### PR TITLE
MRG ENH/BUG Add exclude parameter to plot_evoked_topo

### LINKED ISSUE #9205: ENH/BUG: Adding an include/exclude bad channels parameter to plot_evoked_topo
Was working on `plot_evoked_topo` ,about issue #8714 with @jasmainak when I noticed that bad channels were influencing ylim calculations when no ylim parameters were specified during the function call.  Currently , The plots from `plot_evoked_topo` are automatically set to not plot any of the bad channels and also there is no direct way to force `plot_evoked_topo`  to include the bad channels without removing the channels from `info['bads']` or with picks. 

One thing that we were thinking while fixing the bug was to maybe add a include/exclude parameter to the function to explicitly include or exclude the bad channels in the plot that it produces.  Thoughts? 

@larsoner @agramfort @sappelhoff @drammock 

#### Steps to reproduce
```python
import numpy as np

import mne


from mne.datasets import sample

data_path = sample.data_path()

fname = data_path + '/MEG/sample/sample_audvis-ave.fif'

# Reading evoked data
condition = 'Left Auditory'
evoked = mne.read_evokeds(fname, condition=condition, baseline=(None, 0),
                          proj=False)

evoked.plot_topo()
```

### PR DESCRIPTION
**Pull request by ramkpari**
#### Reference issue
Fixes #9205 .

#### What does this implement/fix?
Adds an exclude parameter to `plot_evoked_topo`.

With the added exclude parameter, ylim's calculations are no longer affected by bad channels that aren't included in the actual plot.

### CODE REVIEW

File: mne/evoked.py [hunk 1]
```diff
@@ -410,21 +410,21 @@ def plot_image(self, picks=None, exclude='bads', unit=True, show=True,
...
 
     @copy_function_doc_to_method_doc(plot_evoked_topo)
-    def plot_topo(self, layout=None, layout_scale=0.945, color=None,
-                  border='none', ylim=None, scalings=None, title=None,
-                  proj=False, vline=[0.0], fig_background=None,
-                  merge_grads=False, legend=True, axes=None,
+    def plot_topo(self, layout=None, layout_scale=0.945,
+                  color=None, border='none', ylim=None, scalings=None,
+                  title=None, proj=False, vline=[0.0], fig_background=None,
+                  merge_grads=False, legend=True, axes=None, exclude=None,
```
Comment (jasmainak) [inline mne/evoked.py:416]: you should add it at the end before `show` so there is minimal chance of breaking existing code

Comment (ramkpari) [reply mne/evoked.py]: Alright, Added that in commit https://github.com/mne-tools/mne-python/pull/9278/commits/3630b0e68bfe91801d375984dc5b4690c2edf2d4

File: mne/viz/evoked.py [hunk 2]
```diff
@@ -840,6 +841,9 @@ def plot_evoked_topo(evoked, layout=None, layout_scale=0.945, color=None,
         Can be a string to load a covariance from disk.
 
         .. versionadded:: 0.16.0
+    exclude : list of str | 'bads'
+        Channels names to exclude from being shown. If 'bads', the
```
Comment (agramfort) [inline mne/viz/evoked.py:845]:
```suggestion
        Channels names to exclude from the plot. If 'bads', the
```

File: mne/viz/topo.py [hunk 3]
```diff
@@ -563,11 +563,12 @@ def _erfimage_imshow_unified(bn, ch_idx, tmin, tmax, vmin, vmax, ylim=None,
...
-                      border='none', ylim=None, scalings=None, title=None,
-                      proj=False, vline=(0.,), hline=(0.,), fig_facecolor='k',
-                      fig_background=None, axis_facecolor='k', font_color='w',
-                      merge_channels=False, legend=True, axes=None, show=True,
+def _plot_evoked_topo(evoked, layout=None, layout_scale=0.945,
+                      color=None, border='none', ylim=None, scalings=None,
+                      title=None, proj=False, vline=(0.,), hline=(0.,),
+                      fig_facecolor='k', fig_background=None,
+                      axis_facecolor='k', font_color='w', merge_channels=False,
+                      legend=True, axes=None, exclude=None, show=True,
```
Comment (agramfort) [inline mne/viz/topo.py:571]:
```suggestion
                      legend=True, axes=None, exclude=('bads,), show=True,
```

File: mne/viz/topo.py [hunk 4]
```diff
@@ -672,21 +676,24 @@ def _plot_evoked_topo(evoked, layout=None, layout_scale=0.945, color=None,
         evoked = [whiten_evoked(e, noise_cov) for e in evoked]
     else:
         evoked = [e.copy() for e in evoked]
+
+    if exclude is None:
+        exclude = ['bads']
+
```
Comment (agramfort) [inline mne/viz/topo.py:682]:
```suggestion

```

**Review by agramfort — COMMENTED** [review]
@drammock do you an easy way to test this?

**Comment by ramkpari** [comment]
Looks like there's some issue with the picks from exclude being reflected. Converting it back into a WIP temporarily

**Comment by ramkpari** [comment]
Also , Somehow that last commit seems to break the entire function

File: mne/viz/topo.py [hunk 5]
```diff
@@ -563,11 +563,12 @@ def _erfimage_imshow_unified(bn, ch_idx, tmin, tmax, vmin, vmax, ylim=None,
...
-                      border='none', ylim=None, scalings=None, title=None,
-                      proj=False, vline=(0.,), hline=(0.,), fig_facecolor='k',
-                      fig_background=None, axis_facecolor='k', font_color='w',
-                      merge_channels=False, legend=True, axes=None, show=True,
+def _plot_evoked_topo(evoked, layout=None, layout_scale=0.945,
+                      color=None, border='none', ylim=None, scalings=None,
+                      title=None, proj=False, vline=(0.,), hline=(0.,),
+                      fig_facecolor='k', fig_background=None,
+                      axis_facecolor='k', font_color='w', merge_channels=False,
+                      legend=True, axes=None, exclude=('bads'), show=True,
```
Comment (drammock) [inline mne/viz/topo.py:571]: Is this meant to be a tuple? If so it's missing a comma inside the parentheses

Comment (ramkpari) [reply mne/viz/topo.py:571]: No. It's supposed to be just a string in this case.

**Comment by ramkpari** [comment]
The plot updating should be fixed. Also, set default param for exclude to 'bads' for all possible calls

File: mne/evoked.py [hunk 6]
```diff
@@ -414,18 +414,20 @@ def plot_topo(self, layout=None, layout_scale=0.945, color=None,
                   border='none', ylim=None, scalings=None, title=None,
                   proj=False, vline=[0.0], fig_background=None,
                   merge_grads=False, legend=True, axes=None,
-                  background_color='w', noise_cov=None, show=True):
+                  background_color='w', noise_cov=None, exclude=('bads'),
```
Comment (agramfort) [inline mne/evoked.py:417]: my bad it should be
```suggestion
                  background_color='w', noise_cov=None, exclude='bads',
```
following the API of https://mne.tools/stable/generated/mne.pick_types.html for example.
Also do not forget to update the docstring. If should be a reflex. When you touch the signature of a public function you must update the docstring accordingly.

**Review by agramfort — COMMENTED** [review]
do you have a tiny code snippet to test / debug this PR? it would allow to test quickly ourselves.

File: mne/viz/topo.py (hunk 5 above)
Comment (ramkpari) [inline mne/viz/topo.py:571]:
```suggestion
                      legend=True, axes=None, exclude='bads', show=True,
```

File: mne/viz/evoked.py [hunk 7]
```diff
@@ -770,11 +770,12 @@ def plot_evoked(evoked, picks=None, exclude='bads', unit=True, show=True,
...
 
-def plot_evoked_topo(evoked, layout=None, layout_scale=0.945, color=None,
-                     border='none', ylim=None, scalings=None, title=None,
-                     proj=False, vline=[0.0], fig_background=None,
+def plot_evoked_topo(evoked, layout=None, layout_scale=0.945,
+                     color=None, border='none', ylim=None, scalings=None,
+                     title=None, proj=False, vline=[0.0], fig_background=None,
                      merge_grads=False, legend=True, axes=None,
-                     background_color='w', noise_cov=None, show=True):
+                     background_color='w', noise_cov=None, exclude=('bads'),
```
Comment (ramkpari) [inline mne/viz/evoked.py:777]:
```suggestion
                     background_color='w', noise_cov=None, exclude='bads',
```

File: mne/evoked.py (hunk 6 above)
Comment (ramkpari) [reply mne/evoked.py]: In this case, I assumed `('bads')` would be equivalent to `'bads'` since a set of one string is still passed as just the string.

**Comment by ramkpari** [comment]
> do you have a tiny code snippet to test / debug this PR? it would allow to test quickly ourselves.

I've been using the following code with some edits for different possible calls.

```python
import numpy as np


import mne
from mne.viz import (plot_evoked_topo)
from mne.datasets import sample
import matplotlib.pyplot as plt
data_path = sample.data_path()

fname = data_path + '/MEG/sample/sample_audvis-ave.fif'

# Reading evoked data
condition = 'Left Auditory'
evoked = mne.read_evokeds(fname, condition=condition, baseline=(None, 0),
                          proj=False)

# Specific channel 
evoked.plot_topo(exclude=['MEG 0113'])

# Default/None
evoked.plot_topo()

# Empty and exclude none
evoked.plot_topo(exclude=[])

# With merge_channels = True
plot_evoked_topo([evoked, evoked], merge_grads=True,
                     background_color='w')

```

File: mne/evoked.py (hunk 6 above)
Comment (drammock) [reply mne/evoked.py]: `'bads'` and `('bads')` are indeed treated the same (but the latter is not a "set", that would be `{'bads'}`).  Removing the parentheses when they're not needed is done for readability (nobody will accidentally think it was a tuple or wonder if it was meant to be a tuple) so it makes the code easier to maintain.

Comment (ramkpari) [reply mne/evoked.py]: Ahh, Fair enough.

File: mne/viz/topo.py [hunk 8]
```diff
@@ -632,12 +633,15 @@ def _plot_evoked_topo(evoked, layout=None, layout_scale=0.945, color=None,
...
         Axes to plot into. If None, axes will be created.
-    show : bool
-        Show figure if True.
     noise_cov : instance of Covariance | str | None
         Noise covariance used to whiten the data while plotting.
         Whitened data channels names are shown in italic.
         Can be a string to load a covariance from disk.
+    exclude : list of str | 'bads'
+        Channels names to exclude from being shown. If 'bads', the
+        bad channels are excluded. If None, exclude is set to 'bads'.
```
Comment (jasmainak) [inline mne/viz/topo.py:642]: if it can be `None`, you should add None also in L640

**Comment by jasmainak** [comment]
Just a thought. Since we now allow showing bad channels with `exclude=[]` I think we should also show it in red like in the other plotting functions. How difficult would it be to add?

File: mne/viz/topo.py (hunk 8 above)
Comment (jasmainak) [reply mne/viz/topo.py:642]: actually I just tried and `None` does not work

**Comment by jasmainak** [comment]
Code looks clean otherwise. Set it MRG when it's ready from your end

File: mne/viz/topo.py (hunk 8 above)
Comment (ramkpari) [reply mne/viz/topo.py:642]: `None` in this case would be the lack of any input param. So basically indicating that exclude would default to 'bads' if no input is received.

Comment (ramkpari) [reply mne/viz/topo.py:642]: Should I update the docstring to refernce it as `By default, it's set to 'bads'` ?

**Comment by ramkpari** [comment]
> Just a thought. Since we now allow showing bad channels with `exclude=[]` I think we should also show it in red like in the other plotting functions. How difficult would it be to add?

Looks like colour is being set [here]( https://github.com/mne-tools/mne-python/blob/main/mne/viz/topo.py#L479-L482) based on the condition it's being plotted.  Since it's being plotted with `unified=True` the colour is being set in one go for an entire condition in those lines. 

Given that it's possible to plot multiple conditions at once with `plot_topo`, I don't think it would fit in with a plot like [this](https://mne.tools/dev/_images/sphx_glr_plot_70_fnirs_processing_022.png)

**Comment by drammock** [comment]
> @drammock do you an easy way to test this?

I (finally) got around to looking into how you might test this.  One simple approach would be to test `len(fig.axes[0].collections)` which should be shorter/longer depending on how many channels you excluded.  No need to mess with `unified=False` that way.

**Comment by jasmainak** [comment]
+1 to add a test based on what @drammock suggests

File: mne/viz/topo.py (hunk 8 above)
Comment (jasmainak) [reply mne/viz/topo.py:642]: yep, update the docstring!

**Comment by ramkpari** [comment]
> > @drammock do you an easy way to test this?
> 
> I (finally) got around to looking into how you might test this. One simple approach would be to test `len(fig.axes[0].collections)` which should be shorter/longer depending on how many channels you excluded. No need to mess with `unified=False` that way.

How does the current test look? . Does it need more test options / simplifying?

**Review by jasmainak — APPROVED** [review]
I didn't test locally on my computer but diff looks great!

File: mne/viz/tests/test_topo.py [hunk 9]
```diff
@@ -181,6 +181,11 @@ def test_plot_topo():
     evoked.pick_types(meg=True).plot_topo(noise_cov=cov)
     plt.close('all')
 
+    # Test exclude parameter
+    fig = picked_evoked.plot_topo(exclude=['MEG 0112'])
+    ch_count = len(picked_evoked.info['ch_names'])
+    assert len(fig.axes[0].collections) - 1 == (ch_count - 1) * 2
```
Comment (drammock) [inline mne/viz/tests/test_topo.py:187]: OK, it turns out that the `LineCollection`s are the horiz/vert axes lines and the background patches; the actual data are in `fig.axes[0].lines`, so we should count that instead. Just to make this super clear for the next person who looks at it, let's assign the two things being compared here to variables with descriptive names:

```suggestion
    exclude = ['MEG 0112']
    fig = picked_evoked.plot_topo(exclude=exclude)
    n_axes_expected = len(picked_evoked.info['ch_names']) - len(exclude)
    n_axes_found = len(fig.axes[0].lines)
    assert n_axes_found == n_axes_expected
```

**Review by drammock — APPROVED** [review]

**Comment by agramfort** [comment]
thx @ramkpari !

**Pull request merged**